.circleci/config.yml

```diff
@@ -53,26 +146,122 @@ unit-tests-defaults: &unit-tests-defaults
   executor: golang
   resource_class: medium+
 
+# The default circle ci checkout does a full checkout and that brings about
+# 450M of history that we don't need.
+shallow-checkout: &shallow-checkout
+  run:
+    name: shallow checkout
+    command: |
+      mkdir ~/.ssh && ssh-keyscan github.com >> ~/.ssh/known_hosts
+      git clone --quiet --depth 1 "$CIRCLE_REPOSITORY_URL" --branch "$CIRCLE_BRANCH" ~/repos/geth
+
+install-go: &install-go
+  run:
+    name: install go
+    command: |
+      # Install go so the e2e tests can build geth
+      mkdir -p ~/repos/golang
+      wget https://dl.google.com/go/go1.16.4.linux-amd64.tar.gz
+      tar xf go1.16.4.linux-amd64.tar.gz -C ~/repos/golang
+      ~/repos/golang/go/bin/go version
+
+save-go-cache: &save-go-cache
+  save_cache:
+    key: go-cache-<<pipeline.parameters.go-cache-version>>-{{ checksum "~/repos/geth/go.sum" }}-{{ checksum "~/repos/month" }}
+    paths:
+      - ~/repos/gocache
+
+restore-go-cache: &restore-go-cache
+  restore_cache:
+    keys:
+      - go-cache-<<pipeline.parameters.go-cache-version>>-{{ checksum "~/repos/geth/go.sum" }}-{{ checksum "~/repos/month" }}
+
+go-mod-cache: &save-go-mod-cache
+  save_cache:
+    key: go-mod-<<pipeline.parameters.go-cache-version>>-{{ checksum "~/repos/geth/go.sum" }}
+    paths:
+      - ~/repos/gocache
+
+restore-go-mod-cache: &restore-go-mod-cache
+  restore_cache:
+    keys:
+      - go-mod-<<pipeline.parameters.go-cache-version>>-{{ checksum "~/repos/geth/go.sum" }}
+
+store-month-in-file: &store-month-in-file
+  # output the current month to a file, we'll use this to rotate the gocache
+  # monthly.  This is because most of what is in the gocache will be related to
+  # the go modules but some of it will be related to the repo code, but we
+  # don't want to rebuild on every commit, so instead we regenerate the cache
+  # monthly and of course if the modules change that will change the cache key.
+  run: date +%m > ~/repos/month
+
+end-to-end-test: &end-to-end-test
+  executor: e2e
+  resource_class: large
+  working_directory: ~/repos
+  parameters:
+    cache-key:
+      type: string
+      default: checkout-monorepo-{{ checksum "~/repos/geth/monorepo_commit" }}-v<<pipeline.parameters.checkout-monorepo-cache-version>>-<<pipeline.parameters.e2e-executor-image>>
```

Comment:
As we talked, this `monorepo_commit` could bring some issues if for example, someone uses master. That will save a cache for the `master` branch, and with new commits, will continue to access the same cached master from before. Same with all the others branches, but probably as the contract versions are tags and not branches, will be something that could happen to master

Reply:
I've addressed this in dbbe58e472df3882939be59e9e3a894e6e882790 `make prepare-system-contracts` will fail if `monorepo_commit` contains a branch.

Reply:
> I've addressed this in [dbbe58e](https://github.com/celo-org/celo-blockchain/commit/dbbe58e472df3882939be59e9e3a894e6e882790) `make prepare-system-contracts` will fail if `monorepo_commit` contains a branch.

Let me know what you think @gastonponti ?